### ISSUE TITLE
Import failing with "Cannot read property '0' of undefined" (round 2)

### ISSUE THREAD

**Issue by lddubeau** [systemjs/systemjs#1700]
This is essentially #1640 that refuses to die. When we fixed #1640, we certainly fixed a scenario that could cause the issue, but it seems there are other cases that are still not handled. 

### Versions

I get the problem with versions 0.20.12 and 0.20.14. I've not tried other versions.

The fact that I can reproduce it with 0.20.12 is significant since that was the release that incorporated the fix for #1640. The failure with 0.20.12 indicates that the fix for #1640 did not handle all scenarios that could cause the problem. :-(

### Symptoms

I get a big fat exception:

```
Error: Cannot read property '0' of undefined
```

It points to `var depLoad = dependencyInstantiations[i];` in the following code:

```
function makeDynamicRequire (loader, key, dependencies, dependencyInstantiations, registry, state, seen) {
  // we can only require from already-known dependencies
  return function (name) {
    for (var i = 0; i < dependencies.length; i++) {
      if (dependencies[i] === name) {
        var depLoad = dependencyInstantiations[i];
```

The little tracing I did shows that `makeDynamicRequire` is sometimes called with `dependencies.length > 0` but `dependencyInstantiations === undefined`.

### Steps to reproduce

<strike>I do not yet have a nice self-contained way to reproduce the issue. I usually want to have a nice self-contained sets of steps to reproduce issues I report but I'm making an exception here because other people may be running into the same issue and we may be able to get to a resolution faster if other people who run into the issue can produce a small reproduction of the problem faster than I can. The last time (#1640) , it took at least a month before I was able to narrow it down to a reproducible case.</strike>

See [below](https://github.com/systemjs/systemjs/issues/1700#issuecomment-313210132) for a way to reproduce the problem.

**Comment by meriturva** [comment]
You are right, i have made same analysis on my private repo and i got same result: 

`dependencyInstantiations === undefined`.

On our environment we are running few tests (locally and also on teamcity agents). If karma loads all test we have issue, if karma just loads a subset of tests we have no error at all.

When we run tests locally error is throw loading always same file.  On CI we have same behaviour but on a different file (really strange) and i repeat: if we load a subset of tests no error!

It is reproducible (every build now fails) but on our private repo, so no way to share code base.

**Comment by lddubeau** [comment]
### How to reproduce the problem

1. Clone https://github.com/lddubeau/systemjs_issue4

2. npm install.

3. Start the server ``./server.js localhost:[some port number].`` **It is important that you use THIS SERVER. The problem here is a race condition, and the server purposely delays responding to the request for one module so as to recreate the conditions necessary for the problem to occur.**

4. Direct your browser to load ``http://localhost:[same port number]/index.html``.

If you are lucky, you'll immediately get:

```
Uncaught (in promise) Error: Cannot read property '0' of undefined
  Loading x
    at register-loader.js:531
    at doEvaluate (register-loader.js:609)
    at ensureEvaluate (register-loader.js:519)
    at register-loader.js:537
    at doEvaluate (register-loader.js:609)
    at ensureEvaluate (register-loader.js:519)
    at register-loader.js:537
    at doEvaluate (register-loader.js:609)
    at ensureEvaluate (register-loader.js:519)
    at register-loader.js:537
```

If you are unlucky, you may need to reload until you get the error, or it might require that you play with the timeout that is set in index.html.

When I try it here I get the error nearly 100% of the time.

The two relevant changes from the earlier test case I produced for #1640 were to add `./e` as a dependency to module `d` and make `server.js` delay the loading of `e` instead of delaying the loading of `d`.

### Observations

Upon examining the code more and instrumenting it to see what is going on, I've got the current working hypothesis.

I am assuming that `ensureEvaluate` must not run until all modules that need evaluation have their load record (`load`) satisfy the test `load.linkRecord.linked === true`. This entails that the same test must be satisfied by the time the function returned by `makeDynamicRequire` runs.

Indeed, I've modified the start of `makeDynamicRequire` to this:

```
function makeDynamicRequire (loader, key, link, dependencies, dependencyInstantiations, registry, state, seen) {
  // we can only require from already-known dependencies
  return function (name) {
    console.log("XXX dynamic require", key, link.linked);
```

And modified its invocation to pass the link in the 3rd argument. Whenever I get the error I reported in my initial bug report, I also get reports on the console where `link.linked` is `false`. Whenever I do not get the error, then there is no report of `link.linked` being `false`.

What's going on? I believe that the suggestion I gave in #1640 only *partially* closed the door to the issue. I had not realized that `depsInstantiatePromise` is set only to a promise that resolves the immediate dependencies, and not the deep dependencies. The modification I had suggested waits only for `depsInstantiatePromise`, which helps, but it does not wait for the deep dependencies, and thus it is *possible* for a promise returned by `instantiateDeps` to resolve before all dependencies have their link record be properly linked.

I see the code aiming at these goals:

1. The promise returned by `instantiateDeps` should resolve only when involved modules have been linked. (This is the problem we're facing: the promise sometimes resolves when some modules are not linked.)

2. A circular dependency must not bring everything to a halt (or crash). I'm assuming that's what the `seen` logic is meant to do.

I've tried a few modifications and found that it is really easy to end up with code that misses one of the two goals. One thing though: it seems to me that the `seen` logic is more aggressive than it needs to be to handle circular dependencies. In 0.20.14 the code that manages `seen` records all nodes seen *on all dependency paths traversed so far*. However, to prevent traversing the graph endlessly (or to prevent setting promises that are mutually dependent on one another, which is equivalent here), we only need to keep track of the nodes we've seen *on the dependency path that took us to the current node* (rather than *all* paths traversed so far). I've tried with the following code and can no longer reproduce the error (either with the code above, or with my complete application, which loads hundreds of modules):

```
.then(function () {
    // now deeply instantiateDeps on each dependencyInstantiation that is a load record
    var deepDepsInstantiatePromises = [];

    for (var i = 0; i < link.dependencies.length; i++) {
      var depLoad = link.dependencyInstantiations[i];
      var depLink = depLoad.linkRecord;

      if (!depLink || depLink.linked)
        continue;

      if (seen.indexOf(depLoad) !== -1)
        continue;

      var newSeen = seen.slice();
      newSeen.push(depLoad);
      deepDepsInstantiatePromises.push(instantiateDeps(loader, depLoad, depLoad.linkRecord, registry, state, newSeen));
    }

    return Promise.all(deepDepsInstantiatePromises);
  })
```

I've included the modification above in the file `system.updated.js` which is part of the repo I link to above.

**Comment by meriturva** [comment]
@guybedford could you please check that issue? we are CI blocked and we have to take a decision here (move to systemjs 0.19) or wait an official fix.

**Comment by guybedford** [comment]
I've put together a fix along these same lines in https://github.com/ModuleLoader/es-module-loader/commit/ebf9ddacc630fc2883517a95f4252a372ed5d0, maintaining one seen array per top-level load instead of splicing along each branch. The good news is it seems to provide a ~10% perf improvement as well.

Thanks again @lddubeau for the detailed analysis.

**Comment by guybedford** [comment]
Released in 0.20.15.
